server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java

```diff
@@ -177,14 +188,53 @@ public void startReplication(ActionListener<Void> listener) {
         cancellableThreads.checkForCancel();
         state.setStage(SegmentReplicationState.Stage.FILE_DIFF);
-        final Store.RecoveryDiff diff = Store.segmentReplicationDiff(checkpointInfo.getMetadataMap(), indexShard.getSegmentMetadataMap());
-        logger.trace(() -> new ParameterizedMessage("Replication diff for checkpoint {} {}", checkpointInfo.getCheckpoint(), diff));
+        Map<String, StoreFileMetadata> activeReaderMetadataMap = indexShard.getSegmentMetadataMap();
+        Map<String, StoreFileMetadata> primaryMetadataMap = checkpointInfo.getMetadataMap();
+
+        Store.RecoveryDiff diff = Store.segmentReplicationDiff(primaryMetadataMap, activeReaderMetadataMap);
+        List<String> unreferencedDiskFiles = Arrays.asList(indexShard.store().directory().listAll())
+            .stream()
+            .filter(
+                name -> !activeReaderMetadataMap.containsKey(name)
+                    && !EXCLUDE_FILES.contains(name)
+                    && !name.startsWith(IndexFileNames.SEGMENTS)
+            )
+            .collect(Collectors.toList());
+
+        Set<String> unreferencedFilesWithSameChecksum = new HashSet<>();
+        Set<String> unreferencedFilesWithChecksumMismatch = new HashSet<>();
+        for (String file : unreferencedDiskFiles) {
+            // If primary does not contain file, ignore it.
+            if (primaryMetadataMap.containsKey(file) == false) {
+                continue;
+            }
+            try (IndexInput indexInput = indexShard.store().directory().openInput(file, IOContext.DEFAULT)) {
+                String checksum = Store.digestToString(CodecUtil.retrieveChecksum(indexInput));
+                if (primaryMetadataMap.get(file).checksum().equals(checksum)) {
+                    unreferencedFilesWithSameChecksum.add(file);
+                } else {
+                    unreferencedFilesWithChecksumMismatch.add(file);
+                }
+            }
+        }
+        final List<StoreFileMetadata> missingFiles = diff.missing.stream()
+            .filter(md -> unreferencedFilesWithSameChecksum.contains(md.name()) == false)
+            .collect(Collectors.toList());
+
+        logger.trace(
+            () -> new ParameterizedMessage(
+                "Replication diff for checkpoint {} {} {}",
+                checkpointInfo.getCheckpoint(),
+                missingFiles,
+                diff.different
+            )
+        );
         /*
          * Segments are immutable. So if the replica has any segments with the same name that differ from the one in the incoming
          * snapshot from source that means the local copy of the segment has been corrupted/changed in some way and we throw an
          * IllegalStateException to fail the shard
          */
-        if (diff.different.isEmpty() == false) {
+        if (diff.different.isEmpty() == false || unreferencedFilesWithChecksumMismatch.isEmpty() == false) {
```

Comment:
I think we can simply delete the conflicting file instead of throwing here.  If we don't then its not going to get cleaned up and this will fail again on retry?

Reply:
I think `OpenSearchCorruptionException` results in replica shard failure, moving shard copy out-of-sync, marking local store as corrupted (and cleaned up in follow up recovery attempt- need to verify this). Logically, we should never reach this state, but consuming conflicting files is better approach towards recovering from previous failures that led to conflicting files.